File: kora-java-openapi-generator-http-client/src/test/java/ru/tinkoff/kora/example/openapi/http/client/HttpClientPetV3Tests.java
package ru.tinkoff.kora.example.openapi.http.client;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.fail;
import static org.mockserver.model.HttpRequest.request;
import static org.mockserver.model.HttpResponse.response;

import io.goodforgod.testcontainers.extensions.ContainerMode;
import io.goodforgod.testcontainers.extensions.mockserver.ConnectionMockServer;
import io.goodforgod.testcontainers.extensions.mockserver.MockServerConnection;
import io.goodforgod.testcontainers.extensions.mockserver.TestcontainersMockServer;
import java.util.List;
import org.jetbrains.annotations.NotNull;
import org.json.JSONArray;
import org.json.JSONObject;
import org.junit.jupiter.api.Test;
import org.mockserver.model.JsonBody;
import ru.tinkoff.kora.example.openapi.petV3.api.PetApi;
import ru.tinkoff.kora.example.openapi.petV3.api.PetApiResponses;
import ru.tinkoff.kora.example.openapi.petV3.model.Category;
import ru.tinkoff.kora.example.openapi.petV3.model.Pet;
import ru.tinkoff.kora.example.openapi.petV3.model.Tag;
import ru.tinkoff.kora.test.extension.junit5.KoraAppTest;
import ru.tinkoff.kora.test.extension.junit5.KoraAppTestConfigModifier;
import ru.tinkoff.kora.test.extension.junit5.KoraConfigModification;
import ru.tinkoff.kora.test.extension.junit5.TestComponent;

@TestcontainersMockServer(mode = ContainerMode.PER_RUN)
@KoraAppTest(Application.class)
class HttpClientPetV3Tests implements KoraAppTestConfigModifier {

    @ConnectionMockServer
    private MockServerConnection mockserverConnection;

    @NotNull
    @Override
    public KoraConfigModification config() {
        return KoraConfigModification
                .ofSystemProperty("HTTP_CLIENT_PET_V2_URL", mockserverConnection.params().uri().toString())
                .withSystemProperty("HTTP_CLIENT_PET_V3_URL", mockserverConnection.params().uri().toString());
    }

    @TestComponent
    private PetApi petApi;

    @Test
    void getRequestSuccess() {
        // given
        var tags = new JSONArray();
        tags.put(new JSONObject()
                .put("id", 1L)
                .put("name", "tag"));
        var responseBody = new JSONObject()
                .put("id", 1L)
                .put("name", "name")
                .put("status", Pet.StatusEnum.AVAILABLE.getValue())
                .put("category", new JSONObject()
                        .put("id", 1L)
                        .put("name", "category"))
                .put("tags", tags);

        mockserverConnection.client().when(request()
                .withMethod("GET")
                .withPath("/v3/pet/{id}")
                .withPathParameter("id", String.valueOf(responseBody.getLong("id"))))
                .respond(response()
                        .withBody(responseBody.toString()));

        // when
        // then
        var response = petApi.getPetById(1L).block();
        if (response instanceof PetApiResponses.GetPetByIdApiResponse.GetPetById200ApiResponse r) {
            assertEquals(responseBody.getLong("id"), r.content().id());
            assertEquals(responseBody.getString("name"), r.content().name());
            assertEquals(responseBody.getString("status"), r.content().status().getValue());
        } else {
            fail("Shouldn't happen");
        }
    }

    @Test
    void postRequestSuccess() {
        // given
        var tags = new JSONArray();
        tags.put(new JSONObject()
                .put("id", 1L)
                .put("name", "tag"));
        var requestBody = new JSONObject()
                .put("id", 1L)
                .put("name", "name")
                .put("status", Pet.StatusEnum.AVAILABLE.getValue())
                .put("category", new JSONObject()
                        .put("id", 1L)
                        .put("name", "category"))
                .put("tags", tags);

        mockserverConnection.client().when(request()
                .withMethod("POST")
                .withPath("/v3/pet")
                .withBody(new JsonBody(requestBody.toString())))
                .respond(response()
                        .withBody(new JsonBody(requestBody.toString())));

        // when
        // then
        var request = new Pet(1L,
                "name",
                new Category(1L, "category"),
                List.of(new Tag(1L, "tag")),
                Pet.StatusEnum.AVAILABLE);
        var response = petApi.addPet(request).block();
        if (response instanceof PetApiResponses.AddPetApiResponse.AddPet200ApiResponse r) {
            assertEquals(request, r.content());
        } else {
            fail("Shouldn't happen");
        }
    }
}

File: examples/kotlin/kora-kotlin-openapi-generator-http-client/src/test/kotlin/io/koraframework/kotlin/example/openapi/http/client/HttpClientPetV3Tests.kt
package io.koraframework.kotlin.example.openapi.http.client

import io.goodforgod.testcontainers.extensions.ContainerMode
import io.goodforgod.testcontainers.extensions.mockserver.ConnectionMockServer
import io.goodforgod.testcontainers.extensions.mockserver.MockServerConnection
import io.goodforgod.testcontainers.extensions.mockserver.TestcontainersMockServer
import org.json.JSONArray
import org.json.JSONObject
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.fail
import org.junit.jupiter.api.Test
import org.mockserver.model.HttpRequest.request
import org.mockserver.model.HttpResponse.response
import org.mockserver.model.JsonBody
import io.koraframework.kotlin.example.openapi.petV3.api.PetApi
import io.koraframework.kotlin.example.openapi.petV3.api.PetApiResponses
import io.koraframework.kotlin.example.openapi.petV3.model.Category
import io.koraframework.kotlin.example.openapi.petV3.model.Pet
import io.koraframework.kotlin.example.openapi.petV3.model.Tag
import io.koraframework.test.extension.junit5.KoraAppTest
import io.koraframework.test.extension.junit5.KoraAppTestConfigModifier
import io.koraframework.test.extension.junit5.KoraConfigModification
import io.koraframework.test.extension.junit5.TestComponent

@TestcontainersMockServer(mode = ContainerMode.PER_RUN)
@KoraAppTest(Application::class)
class HttpClientPetV3Tests : KoraAppTestConfigModifier {

    @ConnectionMockServer
    lateinit var mockserverConnection: MockServerConnection

    @TestComponent
    lateinit var petApi: PetApi

    override fun config(): KoraConfigModification = KoraConfigModification
        .ofSystemProperty("HTTP_CLIENT_PET_V2_URL", mockserverConnection.params().uri().toString())
        .withSystemProperty("HTTP_CLIENT_PET_V3_URL", mockserverConnection.params().uri().toString())

    @Test
    fun getRequestSuccess() {
        // given
        val tags = JSONArray()
        tags.put(JSONObject().put("id", 1L).put("name", "tag"))
        val responseBody = JSONObject()
            .put("id", 1L)
            .put("name", "name")
            .put("status", Pet.StatusEnum.AVAILABLE.value)
            .put("category", JSONObject().put("id", 1L).put("name", "category"))
            .put("tags", tags)

        mockserverConnection.client().`when`(
            request()
                .withMethod("GET")
                .withPath("/v3/pet/{id}")
                .withHeader("X-API-KEY", "MyAuthApiKey")
                .withPathParameter("id", responseBody.getLong("id").toString())
        ).respond(response().withBody(responseBody.toString()))

        // when
        val response = petApi.getPetById(1L)

        // then
        if (response is PetApiResponses.GetPetByIdApiResponse.GetPetById200ApiResponse) {
            assertEquals(responseBody.getLong("id"), response.content.id)
            assertEquals(responseBody.getString("name"), response.content.name)
            assertEquals(responseBody.getString("status"), response.content.status?.value)
        } else {
            fail("Shouldn't happen")
        }
    }

    @Test
    fun postRequestSuccess() {
        // given
        val tags = JSONArray()
        tags.put(JSONObject().put("id", 1L).put("name", "tag"))
        val requestBody = JSONObject()
            .put("id", 1L)
            .put("name", "name")
            .put("status", Pet.StatusEnum.AVAILABLE.value)
            .put("category", JSONObject().put("id", 1L).put("name", "category"))
            .put("tags", tags)

        mockserverConnection.client().`when`(
            request()
                .withMethod("POST")
                .withPath("/v3/pet")
                .withHeader("X-API-KEY", "MyAuthApiKey")
                .withBody(JsonBody(requestBody.toString()))
        ).respond(response().withBody(JsonBody(requestBody.toString())))

        // when
        // the 2.0 generator orders the constructor by optionality, so the arguments are named
        val request = Pet(
            id = 1L,
            name = "name",
            category = Category(1L, "category"),
            tags = listOf(Tag(1L, "tag")),
            status = Pet.StatusEnum.AVAILABLE
        )
        val response = petApi.addPet(request)

        // then
        if (response is PetApiResponses.AddPetApiResponse.AddPet200ApiResponse) {
            assertEquals(request, response.content)
        } else {
            fail("Shouldn't happen")
        }
    }
}
